_posts/2025-12-12-iceberg-in-the-browser.md

```diff
@@ -0,0 +1,119 @@
+---
+layout: post
+title: "Iceberg in the Browser"
+author: "Carlo Piovesan, Tom Ebergen, Gábor Szárnyas"
+thumb: "/images/blog/thumbs/iceberg-in-the-browser.svg"
+image: "/images/blog/thumbs/iceberg-in-the-browser.png"
+excerpt: "DuckDB is the first end-to-end interface to Iceberg REST Catalogs within a browser tab. You can now read and write tables in an Iceberg catalog without needing to manage any infrastructure – directly from your browser!"
+tags: ["deep dive"]
+---
+
+Accessing an Iceberg REST Catalogs is a complex operation that requires quite a few moving parts.
+In this post, we first describe the current state of connecting to Iceberg catalogs.
+Then, we pose the question: could you do simply from a browser tab?
+We then elaborate on the changes required in the DuckDB ecosystem that unlock this capability.
+Finally, we demonstate our new approach that allows you to interact with an Iceberg REST Catalog by simply navigating to a URL – without any setup required.
+
+## Interaction Models for Iceberg Catalogs
+
+![Iceberg analytics today](/images/blog/iceberg-wasm/iceberg-analytics-today-dark.svg){: .darkmode-img }
+![Iceberg analytics today](/images/blog/iceberg-wasm/iceberg-analytics-today-light.svg){: .lightmode-img }
+
+_Iceberg_ is an _Open Table Format,_ which allows you to capture a mutable database table as a set of static files on object storage (such as AWS S3).
+_Iceberg catalogs_ allow you to track and organize Iceberg tables.
+For example, [Iceberg REST Catalogs](https://iceberg.apache.org/rest-catalog-spec/) provide these functionalities through a REST API.
+
+There are two common ways to interact with Iceberg catalogs:
+
+* The *client–server model,* where the compute part of the operation is delegated to a managed infrastructure (such as the cloud). Users can interact with the server by installing a local client or using a lightweight client such as a browser.
+* The *client-is-the-server model,* where the user first installs the relevant libraries, and then performs queries directly on their machine.
+
+Both models have their tradeoffs. In the *client–server model,* clients can be lightweight, and the server is the uniform point of access allowing for potential efficiencies thanks to scale. However, the infrastructure necessitates additional maintenance. In the *client-is-the-server model,* the query latency is lower, users can leverage local compute resources, and integrate with local inputs and external data sources happens at the user's level. However, requiring users to run computation locally means transferring part of the burden on your users (e.g., setting up dependencies).
+
+To fit these catalog models, the implementation of existing Iceberg engines fall into these categories: they can run locally as binaries or an organization can host them in their server infrastructure.
```

Comment:
```suggestion
A majority of Iceberg engine implementations fall into these interaction models: they can either be run locally by or by an organization that hosts them within their server infrastructure.
```

Reply:
On this, I am not sure.
I think it's both more strong to say we are unique, and I also think that it's broadly true that the engine (that is different from the whole end to end beast) has to run somewhere, either locally or in a server.
Unsure what a better wording would be, I'll look it up this again later.

Reply:
updated the comment. Going with "All Iceberg engine implementations" If someone gets mad, we apologize and update our blog post.